Python3

JuiP opened the pull request:
@quozl @pro-panda please review.
Tested works without errors

- [x] Port to python 3

- [x] Fix PyGIWarnings

- [x] Fix Robot mode - broken after port to Gtk+3

- [x] Port telepathy static bindings to TelepathyGLib

- [x] Update Collabwrapper with latest changes in collab test activity

- [x] Port from GObject to GLib

- [ ] Collaboration seems to be broken before the port

       `` PathsActivity.py line 261
          AttributeError: 'PathsActivity' object has no attribute ' _shared_activity' ``
          Will open another PR to fix this

JuiP committed:
Port to Python 3

JuiP committed:
updating collabwrapper

JuiP committed:
More changes python 2 to 3

JuiP committed:
Python3 changes to sprites.py

JuiP committed:
Changes to sprites.py

JuiP committed:
Revert "Changes to sprites.py"

This reverts commit 6d9d09fd5d90d0b863df9a435b2d994a6a73c6ed.

JuiP committed:
Revert "Python3 changes to sprites.py"

This reverts commit 4c1cbd9628a4fa33018025b92d1a047e72455b81.

JuiP committed:
Fix PyGIwarnings and fix other port to GTK+3 errors

JuiP committed:
Port telepathy static binding to TelepathyGLib

JuiP committed:
run activity translation bash

JuiP committed:
remove bash file

chimosky commented on PathsActivity.py:
```diff
@@ -304,12 +304,12 @@ def _list_tubes_reply_cb(self, tubes):
     def _list_tubes_error_cb(self, e):
         """ Log errors. """
-        print('Error: ListTubes() failed: %s', e)
+        print(('Error: ListTubes() failed: %s', e))
```
I'm guessing you used `2to3` and didn't make any changes after that, this line would print both `%s` and the value of `e`.
```python
Error: ListTubes() failed: %s e
```
There's no need for the double parentheses except you were wrapping the placeholder value which you're not here, 
the line should be
```python
print('Error: ListTubes() failed: %s' %(e))
```
or you can use f strings and it'll look like
```python
print(f'Error: ListTubes() failed: {e}')

JuiP replied:
I seemed to have skipped that.... resolved in commit `aa4103e`

chimosky commented on PathsActivity.py:
```diff
@@ -304,12 +304,12 @@ def _list_tubes_reply_cb(self, tubes):
-        print('New tube: ID=%d initator=%d type=%d service=%s params=%r \
-state=%d' % (id, initiator, type, service, params, state))
+        print(('New tube: ID=%d initator=%d type=%d service=%s params=%r \
+state=%d' % (id, initiator, type, service, params, state)))
```
You don't need the extra parentheses here too, you can either leave it the way it is or use f strings.

chimosky commented on utils.py:
```diff
@@ -58,7 +58,7 @@ def json_dump(data):
     ''' Load pixbuf from SVG string '''
     pl = GdkPixbuf.PixbufLoader.new_with_type('svg')
-    pl.write(svg_string)
+    pl.write(bytes(svg_string.encode()))
```
Why the use of `bytes` here after encoding the string?

chimosky replied:
Saw that you fixed this in a different commit.

JuiP committed:
run python setup.py genpot

chimosky commented:
The `trans.sh` script in 703a0a4 shouldn't be added to the commit as it was used to update po files and isn't used by the activity.

chimosky commented:
You could also take the opportunity and port `GObject` to `GLib`.

JuiP committed:
Fix print statements and format specifiers

JuiP committed:
Fix format specifier error

JuiP commented:
@chimosky should I port to `GLib` in the same PR or open another PR?

chimosky commented:
You can add it here.

quozl commented:
> Tested works without errors
>     * [ ]  Collaboration seems to be broken before the port
>       ```
>        `` PathsActivity.py line 261
>           AttributeError: 'PathsActivity' object has no attribute ' _shared_activity' ``
>       ```
> Will open another PR to fix this

That doesn't look like testing without errors.  Please fix this.  I don't think it needs to be in a separate pull request.  The error is common; failure to change to get_shared_activity during a port to GTK 3 during GCI 2016, possibly because of hasty review that didn't include collaboration testing.

JuiP committed:
Port from GObject to GLib

JuiP committed:
Python 3- Fix regressions

JuiP committed:
Collaboration testing in progress

JuiP committed:
Removing commented code

JuiP commented:
> That doesn't look like testing without errors. Please fix this. I don't think it needs to be in a separate pull request. The error is common; failure to change to get_shared_activity during a port to GTK 3 during GCI 2016, possibly because of hasty review that didn't include collaboration testing.

It seemed to me like the previous port to collabwrapper was not completed and merged hastily. I have made the required changes in another branch `Collaboration`, which I have later merged in this PR.
- Tested the changes on two VMs connected over a virtual network. The logs are clean, however the collaboration is still broken. 
- The shared activity does not share the same screen (activity grid). Maybe I should raise an issue for this?

- [x] Port from GObject to GLib

Apart from that I added few new translations at https://translate.sugarlabs.org/projects/Paths/ a while ago, when I refreshed pot file and run the translation bash earlier, many translations were not fetched. What have I done wrong?

Please test and review. @quozl @chimosky

quozl commented:
Thanks for the update.

> * Tested the changes on two VMs connected over a virtual network. The logs are clean, however the collaboration is still broken.
> * The shared activity does not share the same screen (activity grid). Maybe I should raise an issue for this?

An issue is for recording a problem that you are not going to fix.  It is an invitation to others to fix it.  For problems you are going to fix, just say so in the pull request and we'll wait until you're ready with a fix.  When you need help to fix a problem while in the middle of a pull request, the best place to ask is the pull request (for context), and the second best place is sugar-devel@ mailing list (for wider audience).

It isn't clear what you plan to do; work on a fix, or give up the pull request for the time being until someone else has a fix.

If you are not sure about how collaboration works, look at source code of other activities that successfully collaborate.  Maze is a good but complex example; using TextChannelWrapper but not CollabWrapper.  Memorize is a good example using CollabWrapper.  Level is another good example that goes a step further to use CollabWrapper with a TCP/IP UDP broadcast.

> Apart from that I added few new translations at https://translate.sugarlabs.org/projects/Paths/ a while ago, when I refreshed pot file and run the translation bash earlier, many translations were not fetched. What have I done wrong?

You haven't finished updating the translations.  I don't know what you did, so I can't critique your process.  When I used [my process](https://github.com/sugarlabs/sugar-tools/blob/master/activity-translations) on top of your branch here, I pushed the result as 72e9f4f ("translate.sugarlabs.org synchronisation").  My changes demonstrate what you hadn't updated.  Your new translations are seen, thanks!

chimosky commented on genpieces.py:
```diff
@@ -212,7 +212,7 @@ def generate_tile_2_lines(a, b, c, d, e, f, g, h, scale=1,
 
 def open_file(datapath, filename):
-    return file(os.path.join(datapath, filename), "w")
+    return open(os.path.join(datapath, filename), "w")
```
You can use `open` without importing it from `io`.

JuiP replied:
I agree that open() is a built-in in both python 2 and python 3. In python 2, io.open() was different from open(), however the differences have been removed in python 3. I added that just out of habit. Thanks for noticing, I will remove it in the next commit.

JuiP replied:
Done in commit `766b774`

JuiP commented:
> It isn't clear what you plan to do; work on a fix, or give up the pull request for the time being until someone else has a fix.

I don't think fixing collaboration under the PR - port to python 3 is a good idea, since it was broken even before the port. I planned to club changes related to porting, for reference later(if required). However, I do understand that making changes ultimately needs to make the activity usable and the order of changes doesn't matter much. If you insist I could try fixing the broken collaboration under this PR. Maybe could you possibly test the changes made till the latest commit (`
7065fcc`)?

> You haven't finished updating the translations. I don't know what you did, so I can't critique your process.

I did use the same script you used to update translations. Let me elaborate:
Saved the script as trans.sh, made the script executable with command chmod +x trans.sh and then run the script using ./trans.sh then run python setup.py genpot. Pushed this as commit `
703a0a4` It is important for me to know what could have possibly been a mistake as I plan to add translations to other projects too.

quozl commented:
Thanks.  It's easy to change the title of the pull request, if that's the only doubt you have.  Otherwise what is the procedural problem that would be solved by using a different pull request?  Is it that you don't want to try fixing the collaboration?  If so, I'm fine with that, and we can leave the pull request unfinished or merge it.  We just need to know what you plan.  I'll see if I can find time to test.  I'm not familiar with the activity; do you have or know of a test plan?

Regarding translations, I've no idea why your results were different to mine, but I agree it is important to find out why.  Your commit https://github.com/sugarlabs/paths/pull/8/commits/703a0a40d6de9d9eac72071fd0dfbb23ee4634ca when viewed shows only removal of some po/*.po files, and addition of trans.sh file; no changes to po/ach.po.  Yet my commit 72e9f4f5b09fdbc75aff55842ba46fbd1da0cb08 shows changes to po/ach.po and a list of other files.  My guess is that you didn't complete a `git add` before `git commit`.  You might test that theory by checking your bash history, any diary you may have, or by making a new branch from a commit before 72e9f4f5b09fdbc75aff55842ba46fbd1da0cb08 and repeating the script and `git add`, `git commit`, and then comparing the commits with `git diff`.

chimosky commented:
> I did use the same script you used to update translations. Let me elaborate:
Saved the script as trans.sh, made the script executable with command chmod +x trans.sh and then run the script using ./trans.sh then run python setup.py genpot. Pushed this as commit 703a0a4 It is important for me to know what could have possibly been a mistake as I plan to add translations to other projects too.

The script requires that you pass an argument to it and the argument is the name of the project as listed on translate.sugarlabs.org and running the script without an argument would run the script but only remove empty translations without pulling in any translations.

JuiP commented:
> Otherwise what is the procedural problem that would be solved by using a different pull request? Is it that you don't want to try fixing the collaboration?

Initially, I didnot plan to fix the collaboration due to tight schedules. However, I have made some progress in fixing it, so I will add more commits if I get it working after testing, might take me some time; will finish it as soon as I can.

> I'll see if I can find time to test. I'm not familiar with the activity; do you have or know of a test plan?

I started with playing a game and noted-

- [x] New game can be started
- [x] Robot mode works, with correct changes in the playbutton displaying if it is my turn
- [x] Stop button works as expected
- [x] Using the activity from both the journal and home view restores the initial game state
- [x] Score is calculated after completing a path.
- [x] When out of moves, the game is over.
- [x] Collaboration- disables the robot mode for the sharer (with changes in tool-tips)
- [x] Collaboration- disables the new game and robot mode for the joiner.
- [x] Collaboration- seen as a collaboration instance in the journal with the icon of the joiner
- [x] Collaboration does not share the same game board and the deck of tiles. Changes in one game board are not reflected in the other.
The above actions have no errors in the logs for me.

JuiP commented:
> The script requires that you pass an argument to it and the argument is the name of the project as listed on translate.sugarlabs.org and running the script without an argument would run the script but only remove empty translations without pulling in any translations.

@chimosky That's right, it was such a stupid mistake! I seemed to have overlooked that! Thanks a lot :)

quozl commented:
Thanks!  Added new issue https://github.com/sugarlabs/sugar-tools/issues/3.

JuiP committed:
WIP collab

JuiP committed:
Fix set_player_on_toolbar error and sharer error

JuiP committed:
Sharer starts the new game and makes the first move

JuiP committed:
Fixes #9 ; calls set_sensitive

JuiP commented:
@quozl @chimosky Update on this PR:
I tested on two virtual machines, collaboration works exactly as expected.

>Otherwise what is the procedural problem that would be solved by using a different pull request?

I just wanted to avoid it in this PR, because the work is a little messy. I am not sure if reviewing code commit wise will be of any help. Maybe it would have been easier to review directly as `files changed`, had it been done in another PR. 

Many changes which have been done commit 38e8b8b onwards may be kept or undone in further commits. I found making changes and simultaneously testing them, more better to work with as a result there are multiple commits. (And also I made the mistake of merging a branch in the branch, might make reviewing code more difficult.... sorry for too many commits and the mess) 

Please test the changes made till 9e5e3fe so I can go ahead with
- Using logging.debug instead of print
- flake8 fixes
- and finally a refreshed pot file with fetching translations

quozl commented:
What you describe is a common hesitation, and I'm happy to explain the process further.  I'll quote selectively.

First, remember that a pull request comes from a branch, and anything you do to the branch will be reflected in the pull request.

A branch is made up of a series of commits.

The commits describe your changes, but are independent of the changes.

> > Otherwise what is the procedural problem that would be solved by using a different pull request?
> 
> I just wanted to avoid it in this PR, because the work is a little messy. I am not sure if reviewing code commit wise will be of any help. Maybe it would have been easier to review directly as `files changed`, had it been done in another PR.

Second, remember that _Files changed_ in the pull request is the equivalent of `git diff master`, assuming you haven't moved _master_ to point to another commit since you began the branch. Otherwise it is the aggregate change since the branch began.

> Many changes which have been done (in) commit [38e8b8b](https://github.com/sugarlabs/paths/commit/38e8b8b073086730030adcd437f3fd715d91798a) onwards may be kept or undone in further commits. I found making changes and simultaneously testing them, more better to work with as a result there are multiple commits. (And also I made the mistake of merging a branch in the branch, might make reviewing code more difficult.... sorry for too many commits and the mess)

Third, making changes and immediately testing them is the best way to work, because (a) you get to reflect on the change during the commit message, (b) you can use git to transport the change to a test system, (c) we get to see your thought process and trials (increased trust), and (d) you can easily revert a failed trial change using `git revert` or `git reset`.

What you may be missing is practice in rewriting commits.  This where you don't change the outcome; the files, but you do change the way that git reaches that outcome.

I can't be specific about the tools you should use to rewrite commits, because I don't know what tools you are using to write them in the first place.  But the tools do eventually call git commands, and these commands are what do the work.

The tools that I use are _emacs_ with _magit_ and _git_.  With reference to this [Git Cheatsheat](http://www.ndpsoftware.com/git-cheatsheet.html#loc=workspace;) ... here's my workflow for writing a commit;

* a source file is edited with _emacs_, and saved into the workspace,
* _magit_ is invoked and the change relative to index is reviewed in _diff_ format,
* the whole change, or tiny parts of it are added to the index using _magit_ keystrokes,
* _magit_ is commanded to make a commit from the index.

Most people just use `git add`, or at worst `git add -a`.  What I'm doing above is similar to `git add --interactive`.

For rewriting commits, I've several different workflows, and I choose one based on the nature of the historical error that I wish to make.

* `git log -1` helps me remember the state of the branch in case I need to go back, (also available with `git reflog` if I lose that from scrollback),
* `git rebase -i` if I want to change the order of commits, squash some commits into other commits, or rewrite their messages (like `git commit --amend` for very old commits),
* `git reset` if I want to get my hands dirty and create a new series of commits from the changes.

After the commit history is tidied up, what I do next depends on whether I've had review comments that address specific commits.  For the most part, I `git push --force`, and leave it to GitHub to show the new history.  But when there has been lengthy discussion on specific commits, I'll stow the rewritten history in a new branch, restore my pull request branch to what it was, revert everything in one commit, and then merge in the new branch.  So the pull request has;

* old commits, some of which aren't to be kept,
* a revert of all the old commits,
* a rewrite using new commits.

Without any change to _Files changed_.

Looking at your branch at present, if it were mine, I would try to rewrite and reorder commits to follow this sequence;
* Port to Python 3,
* Update CollabWrapper,
* Fix PyGIWarnings,
* Port telepathy static binding to TelepathyGLib,
* Port from GObject to GLib,
* Fix collaboration,
* Fix format specifiers, genpot, and translations,
* Anything else.

Remember you can't easily lose anything if you remember the git hash somewhere.

Hope that helps.

quozl commented:
Tested 9e5e3feb0a37241755f0689b90b27129cbdff1c3.

* [ ] the toolbar suffers vertical expansion; caused by the label before the help button; the label should be allowed to be wider,
* [ ] the canvas has a grey area which covers a tile and the top border of the playing board,
* [ ] the tiles can only be moved or rotated by clicking the top left corner of the tiles; not sure if that is intended,

I've not tested collaboration fully.  I need to win a game first.  :grin:  I've never seen this activity before.

![Screenshot_udd_2020-04-06_17:16:54](https://user-images.githubusercontent.com/2936481/78532360-7085d500-782a-11ea-8a91-bac9fe36803c.png)

quozl commented:
Quick check of logs; found this but not sure how to cause;

```log
Traceback (most recent call last):
  File "/usr/share/sugar/activities/Paths.activity/PathsActivity.py", line 306, in _message_cb
    self._processing_methods[command][0](payload)
  File "/usr/share/sugar/activities/Paths.activity/PathsActivity.py", line 370, in _play_a_piece
    if self._game.hands[self._game.whos_turn].hand[i] is not None \
IndexError: list index out of range
```

JuiP commented:
> - [ ] the toolbar suffers vertical expansion; caused by the label before the help button; the label should be allowed to be wider,

will get this fixed

> - [ ] the canvas has a grey area which covers a tile and the top border of the playing board,

Issue #7 , will try fixing it!

> - [ ] the tiles can only be moved or rotated by clicking the top left corner of the tiles; not sure if that is intended,

I can't reproduce this. For me the tiles are rotating by clicking on them multiple times and not necessarily only in the top left corner.

> I need to win a game first. :grin: I've never seen this activity before.

Oh you just need to maximize your score by making a closed path and by using as many tiles as possible for the closed path. I actually kind of found this activity fun ! :grinning:

> Quick check of logs; found this but not sure how to cause;

Regarding this error, I didn't get this error initially. Can you confirm if you have tried to restore a shared activity and continued to play to find this error? I'm not sure if there is any other way to reproduce this.
As far as what I noticed: if I attempt to rejoin a shared activity whose sharing was not initiated by me, then the activity restores but now I become the initiator. So now it is my turn (the initiator plays first), after each player plays one chance, the activity is stuck and returns this error. To fix this maybe I have to ensure that I'm not considered an initiator for this activity after restoring. Is this how you got that error?

chimosky commented on PathsActivity.py:
```diff
@@ -387,14 +385,15 @@ def _take_a_turn(self, nick):
             self.collab.post(dict(
-                command=command,
-                payload=payload
+                action = command,
+                new_text = payload
```
I'm guessing you didn't notice, these keyword arguments shouldn't be spaced.

JuiP replied:
I have corrected it in commit `766b774`. I'm not sure what functional change does this make? Or is it to make code uniform across all activities?

quozl replied:
Hi.  It is not a functional change, but a style change.  Style we use is to have no whitespace either side of = in keyboard arguments.  This helps us to immediately recognise it as not a variable assignment.  It showed up when we looked at the diff.

p.s. when you quote commits we can't hover or click on them, please don't wrap them in tick marks.  766b774.

JuiP replied:
Oops, okay! :)

chimosky commented:
Tested 9e5e3fe.

- [ ] When play button is clicked while player hasn't played, the toolbar expands and I think it's as a result of the length of this text that shows beside the button `There are errors -it is still your turn`.

Tested collaboration, sharer joins but game state is yet to be shared.

JuiP commented:
> Tested collaboration, sharer joins but game state is yet to be shared.

Game state is not shared unless the initiator starts a new game.

chimosky commented:
> > Tested collaboration, sharer joins but game state is yet to be shared.
> 
> Game state is not shared unless the initiator starts a new game.

Game state still isn't shared, also noticed that for the joiner the new game button tooltip says they can't start a game but the button still works.

JuiP commented:
> Collaboration still isn't shared

![Screenshot from 2020-04-06 21-07-25](https://user-images.githubusercontent.com/46154968/78576687-a76be300-784a-11ea-8d32-b767642b166a.png)

Changes in one game are reflected in the other and they have shared game boards.... Have I misunderstood by what you meant by collaboration is'nt shared?

> noticed that for the joiner the new game button tooltip says they can't start a game but the button still works.

Oh I know this problem still exists, I had tried `self._new_game_button.set_sensitive(False)` after line 280 in PathsActivity.py as a result when the sharer started a new game, the joiner could not refresh and share the gameboard as the sharer; it couldnot connect to the event new_game() .... still working on it

JuiP committed:
Fix Toolbar Expansion

chimosky commented:
> Changes in one game are reflected in the other and they have shared game boards.... Have I misunderstood by what you meant by collaboration is'nt shared?

No you didn't misunderstand, I've edited my comment.

@quozl kindly test collaboration as game state isn't shared for me.

srevinsaju commented on collabwrapper/test/activity.py:
```diff
@@ -0,0 +1,249 @@
+
+        self._collab.setup()
+        self._last_buddy = None
+
+    def _message_cb(self, collab, buddy, msg):
+
+        def say(string):
+            self._say('%.6f recv from %s@%s ' %
+                      (time.time(), buddy.props.nick, buddy.props.ip4_address))
+            self._say(string)
+
+        self._last_buddy = buddy
+
+        action = msg.get('action')
+        if action == 'echo-reply':
+            text = msg.get('text')
+            latency = (time.time() - float(text)) * 1000.0
+            say('%s %r latency=%.3f ms\n' % (action, text, latency))
+            return
+        if action == 'echo-request':
+            text = msg.get('text')
+            self._collab.post({'action': 'echo-reply', 'text': text})
+            say('%s %r\n' % (action, text))
+            return
+        if action == 'chat':
+            text = msg.get('text')
+            say('%s %r\n' % (action, text))
+            return
+        say('%s\n' % (action))
+
+    def _make_automatic_restart(self):
+        ct = os.stat('activity.py').st_ctime
+
+        def restarter():
+            if os.stat('activity.py').st_ctime != ct:
+                GLib.timeout_add(233, self.close)
+                logging.error('-- restart --')
+                return False
+            return True
+        GLib.timeout_add(233, restarter)
```
Remove this file

JuiP replied:
Done in commit `766b774`

srevinsaju commented on collabwrapper/test/activity/activity-collabwrappertest.svg:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" ?><!-- Generator: Adobe Illustrator 13.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 14948)  --><!DOCTYPE svg  PUBLIC '-//W3C//DTD SVG 1.1//EN'  'http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd' [
+	<!ENTITY stroke_color "#666666">
+	<!ENTITY fill_color "#ffffff">
+]><svg enable-background="new 0 0 55 55" height="55px" id="svg2885" inkscape:version="0.47 r22583" sodipodi:docname="object-rotate-left.svg" style="" version="1.1" viewBox="0 0 55 55" width="55px" x="0px" xml:space="preserve" xmlns="http://www.w3.org/2000/svg" xmlns:cc="http://creativecommons.org/ns#" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:inkscape="http://www.inkscape.org/namespaces/i ...
+	<inkscape:perspective id="perspective3064" inkscape:persp3d-origin="0.5 : 0.33333333 : 1" inkscape:vp_x="0 : 0.5 : 1" inkscape:vp_y="0 : 1000 : 0" inkscape:vp_z="1 : 0.5 : 1" sodipodi:type="inkscape:persp3d" style=""/></defs><sodipodi:namedview bordercolor="#666666" borderopacity="1" gridtolerance="10" guidetolerance="10" id="namedview2894" inkscape:current-layer="svg2885" inkscape:cx="20.225809 ...
+
+
+
+<g id="g3852" style="" transform="translate(0.0700615,0)"><polygon id="polygon2890" points="12.183,32.074 7.513,23.012 21.07,27.354 " style="fill:&stroke_color;;stroke:&stroke_color;;stroke-linejoin:round" transform="matrix(0.31105043,0.9503934,-0.9503934,0.31105043,36.036712,6.5316011)"/><polygon id="polygon2890-6" points="7.513,23.012 21.07,27.354 12.183,32.074 " style="fill:&stroke_color;;stro ...
```
Remove this file too

JuiP replied:
Done in commit `766b774`

srevinsaju commented on collabwrapper/test/activity/activity.info:
```diff
@@ -0,0 +1,9 @@
+[Activity]
+name = CollabWrapperTest
+activity_version = 1
+bundle_id = org.sugarlabs.CollabWrapperTest
+exec = sugar-activity3 activity.CollabWrapperTestActivity
+icon = activity-collabwrappertest
+license = GPLv2+
+repository = https://github.com/sugarlabs/collabwrapper
+summary = Test CollabWrapper
```
Lol, This too. These are actually test files to check collaboration

JuiP replied:
Done in commit `766b774`

srevinsaju commented on collabwrapper/test/setup.py:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python
+
+# Copyright (C) 2009, Simon Schampijer
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+
+from sugar3.activity import bundlebuilder
+
+bundlebuilder.start()
```
This

JuiP replied:
Done in commit `766b774`

srevinsaju commented on collabwrapper/texteditor.py:
```diff
@@ -0,0 +1,207 @@
+    def __text_buffer_inserted_cb(self, textbuffer, start, text, length):
+        if self._callbacks_status is False:
+            return
+        if self.has_initialized is False:
+            self.has_initialized = True
+        logging.debug('Text inserted is %s' % (text))
+        logging.debug('Text has been updated, %s' % (textbuffer.get_text(
+            textbuffer.get_start_iter(), textbuffer.get_end_iter(), True)))
+        self._collab.post(dict(action='entry_inserted',
+                               res_id=self._id,
+                               start_iter_offset=start.get_line_offset(),
+                               start_iter_line=start.get_line(),
+                               new_text=text))
+
+    '''
+    The text buffer deleted callback is called whenever any text is
+    removed in the editor, so that other users get updated with
+    these changes.
+
+    Args:
+        textbuffer (:class:`Gtk.TextBuffer`): text storage widget
+        start (:class:`Gtk.Iterator`): a pointer to the start position
+        end (:class:`Gtk.Iterator`): a pointer to the end position
+    '''
+
+    def __text_buffer_deleted_cb(self, textbuffer, start, end):
+        if self._callbacks_status is False:
+            return
+        if self.has_initialized is False:
+            self.has_initialized = True
+        logging.debug('Text deleted is %s' %
+                      (textbuffer.get_text(start, end, True)))
+        logging.debug('Text has been updated, %s' % (textbuffer.get_text(
+            textbuffer.get_start_iter(), textbuffer.get_end_iter(), True)))
+        self._collab.post(dict(action='entry_deleted',
+                               res_id=self._id,
+                               start_iter_offset=start.get_line_offset(),
+                               start_iter_line=start.get_line(),
+                               end_iter_offset=end.get_line_offset(),
+                               end_iter_line=end.get_line()))
```
This one too

JuiP replied:
Done in commit `766b774`

srevinsaju commented on toolbar_utils.py:
```diff
@@ -106,7 +106,7 @@ def radio_factory(name, toolbar, callback, cb_arg=None, tooltip=None,
     ''' Factory for adding a label to a toolbar '''
     label = Gtk.Label(label_text)
-    label.set_line_wrap(True)
+    label.set_line_wrap(False)
```
Can you explain me the change necessary here?

quozl replied:
This change did fix the toolbar vertical expansion problem.  It may risk horizontal expansion instead on small screens.  :grin:

srevinsaju replied:
Nice :joy:

JuiP replied:
I reduced the length of the text, so hopefully there won't be an issue of horizontal expansion. Done in commit `766b774`

srevinsaju reviewed (changes requested):
Hey, I am back reviewing.

Quick file diff review and testing.
* [x] Tiles do not fit screen resolution
![image](https://user-images.githubusercontent.com/48695438/78635267-8341ed80-78ae-11ea-88ca-3ecc129c759e.png)

Notice that the top tile is hidden under the toolbar
* [ ] Pressing the help button changes the orientation of the first tile in order ( I guess that's done on mistake)
* [ ] Add help to the help button. Otherwise, the help button might not be working after / before the port
* [ ] Change the tooltip of "Play with Robot" alternatively. i.e A student might not understand if a robot with open eyes means the Robot is playing, but a robot with closed eyes means a robot is not playing. Most of the time, a person new to the activity is most likely to hover over the button and see "Play with the robot" as the default tooltip. When the robot is enabled, the tooltip should change to "Disable Robot", etc.
* [x] Why is the XO icon on the PathsActivity? I am curious


Tested on Arch Linux, Sugar 0.116

quozl commented:
Thanks.  Reviewed.  In addition to comments, I've already made ...

* [x] don't include the CollabWrapper test activity, it isn't needed in this activity,
* [x] don't include the CollabWrapper text editor, it isn't needed here either,
* [x] the accelerator for the StopButton is not needed, as the toolkit does it,
* [ ] there are two different joined signals, from different software layers, but you've linked them to the same callback; see the test activity for how they are different,
* [x] there's a use of incorrect GTK signal name; key_press_event is now key-press-event, and underscores aren't the convention we've used elsewhere,

Tested in single-user mode.

* [ ] clicking on black parts of a card is ignored,

srevinsaju commented:
@quozl is the XO icon kept on purpose in the paths activity?

quozl commented:
@srevinsaju, it is used to show whose turn it is when collaborating.  Colour and tooltip text changes.  Have you got collaboration working yet on your setup?

@JuiP, something is wrong with collaboration, but I'm not sure what.  Grid is shared.  Card hands are unique, as they ought to be. When I play cards to the grid, sometimes the turn is not ended, sometimes it is.  It can get into a state where nobody can play their cards.  I'm not sure what the play and stop button is for in the toolbar.  At one stage, after winning a path, pressing the play bar incremented the score repeatedly.

srevinsaju commented:
> Colour and tooltip text changes. Have you got collaboration working yet on your setup?

No, I haven't tried collaboration yet. Thanks for the info on that. Maybe @JuiP can add that to the help button when it is added / fixed

quozl commented:
The "IndexError: list index out of range" continues to occur in logs.  See [traceback](https://github.com/sugarlabs/paths/pull/8#issuecomment-609610784).

> Can you confirm if you have tried to restore a shared activity and continued to play to find this error?

No.  A new instance was started (right-click, start new).  It was shared.  It was joined by the other user. The traceback was in the sharer's logs.

quozl commented:
At present the help button suggests a move.  It doesn't provide help beyond that.

srevinsaju commented:
But I am unable to see the help window. Is it working for you @quozl ?

JuiP commented:
> But I am unable to see the help window. Is it working for you @quozl ?

There is no help window, help but only shows the positions in which you can add another tile to continue.

> Tiles do not fit screen resolution

Any hints on how to get this fixed? I tried but couldn't find what may have caused this. Also saw many activities have similar issues, is there any activity for which such an issue was fixed?

> No. A new instance was started (right-click, start new). It was shared. It was joined by the other user. The traceback was in the sharer's logs.

That's surprising because I don't get this error in logs. What is even more intriguing is why is collaboration working differently for different people! 

> At one stage, after winning a path, pressing the play bar incremented the score repeatedly.

I have seen this but haven't figured the cause yet. This activity isn't maintained / updated for years and it requires a lot of work and fixes. :worried:

srevinsaju commented:
@JuiP 
> Any hints on how to get this fixed? I tried but couldn't find what may have caused this. Also saw many activities have similar issues, is there any activity for which such an issue was fixed?

Try `Gdk.Screen.height() - GRID_CELL_SIZE`

quozl commented:
> > Tiles do not fit screen resolution
> 
> Any hints on how to get this fixed? I tried but couldn't find what may have caused this. Also saw many activities have similar issues, is there any activity for which such an issue was fixed?

Underlying cause is that most of the activities were written for the OLPC XO with 1200x900 display resolution and 4:3 aspect ratio.

Nowadays it is almost impossible to order large volumes of 4:3 aspect ratio display modules at a reasonable price per pixel, and most modules have an aspect ratio closer to 16:9.

Sugar is used on different systems than before, and the systems have different display resolutions.

So activities have to be changed to adapt.

Looking at the source code;
* the activity canvas is a Gtk.DrawingArea with a set_size_request matching the full display resolution; wishful coding,
* the toolkits will negotiate this down to height minus the toolbar height,
* there's no detection of resolution change in the Activity instance, (resolution changes on display rotation or when attaching an external video sink),
* the canvas is handed to Game,
* there's no detection of size change in the Game instance,
* the draw callback incorrectly uses the canvas allocation to position a rectangle and clip; this would only work if there was no toolbar.

So my guess is that the port from GTK 2 to GTK 3 is the point at which the problem was introduced.

Have a look at other draw callbacks in activities that use sprites.py to see how it could have been solved.  Also compare sprites.py carefully against other activities, to the point of being able to explain the nature of the differences to yourself.  This may show other causes.

srevinsaju commented:
![image](https://user-images.githubusercontent.com/48695438/78750378-c7012980-7978-11ea-9979-9b44fbcacb2e.png)
4:3

![image](https://user-images.githubusercontent.com/48695438/78750478-fe6fd600-7978-11ea-8497-cee34c165d70.png)
16:9
The fix is very unusual. Idk why it works, I found it out after quite lot of debugging :joy:. WIll create a PR shortly

srevinsaju commented:
@JuiP Created PR on https://github.com/JuiP/paths/pull/2.
Please review

quozl commented:
Yes, it is unusual.  The thing is, the Cairo Context passed to the draw signal by GTK already has a [clip rectangle that matches the dimensions of the screen area exposed](https://lazka.github.io/pgi-docs/Gtk-3.0/classes/Widget.html#Gtk.Widget.signals.draw), and there's no need in other activities to create a _new_ clip rectangle.  It was common in GTK 2 with the expose signal.

quozl committed:
Fixes #7

JuiP commented:
I found just changing the size of the rectangle fixed it didn't really need to make changes to the canvas height and change order of statements! @srevinsaju Thanks for helping! :)

srevinsaju commented:
@JuiP , please maintain authorship in the commit https://github.com/sugarlabs/paths/pull/8/commits/8b2e74c52c227e2778576582b825dd5689287999, Thanks. Also Include my commit message to make it helpful for future developers to understand what actually happened amid commits

> I found just changing the size of the rectangle fixed it didn't really need to make changes to the canvas height and change order of statements! @srevinsaju Thanks for helping! :)

The others code changes. For example autorotate and potrait activities for potrait devices. The current code won't help. There are chances for overflow of tiles in potrait activities. As a precautionary measure, I had added the commit, and are just for conventional changes. The code is used because it has been used in other activities. @quozl yoy may also like to see https://github.com/JuiP/paths/pull/2

quozl commented:
I've said it before, and I'll say it again until you hear me; there is **no need** to replace the clip rectangle of the Cairo context.  The code is a holdover from GTK 2.

Relative to 8fc47dd, this works fine;

```diff
diff --git a/game.py b/game.py
index 36aa722..d2b1708 100644
--- a/game.py
+++ b/game.py
@@ -772,13 +772,6 @@ class Game():
         return True
 
     def do_draw(self, cr):
-        ''' Handle the expose-event by drawing '''
-        # Restrict Cairo to the exposed area
-        alloc = self._canvas.get_allocation()
-
-        cr.rectangle(alloc.x, alloc.y - GRID_CELL_SIZE*2, alloc.width, alloc.height)
-        cr.clip()
-        # Refresh sprite list
         self._sprites.redraw_sprites(cr=cr)
 
     def _destroy_cb(self, win, event):
```

Added benefit of removing some out of date comments (there's no expose event in GTK 3), and useless comments.

quozl commented:
Tested further.  There is padding below the grid but not above, and this fixes it for me;

```diff
@@ -33,5 +33,5 @@ class Grid:
         self.left = int((width - (tile_width * COL)) // 2 + tile_width)
         self.xinc = int(tile_width)
-        self.top = 0
+        self.top = 1
         self.yinc = int(tile_height)
 
```

JuiP committed:
Other Fixes

- removed collab test activity
- use open without importing from io
- fix keyword spacing
- reduce the length of labels to prevent horizontal toolbar expansion
- change tooltip to "Disable Robot"
- remove stop button accelerator
- correct the incorrect GTK signal name to "key-press-event"

quozl commented:
Thanks.  Good progress.  Reviewed as at 766b774.

* [ ] don't include the CollabWrapper files .gitignore, README or .flake8, as these have no purpose here,
* [ ] in PathsActivity.py you have changed to importing TelepathyGLib, ExportedGObject and TubeConnection, but none of these three imports are used; always check if an import is used before porting it, and if it is not used remove it,
* [ ] in `_shared_cb` and `_joined_cb` you assign a value to a local variable that is not used within the method,
* [ ] your reversal of the conditional check in `_game_over` is odd; are you sure about that?
* [ ] your call to `set_player_on_toolbar` is missing a self, are you sure it works?
* [ ] in `activity/activity.info` git shows you have fixed the missing end of line mark at the end of file, but the repository is still wrong; good opportunity to validated and fix it.  (you may not have intentionally fixed the missing end of line mark; it could have been your source code editor),

After fixing these, there might be other things I'll notice.  I'll also do another test.

JuiP commented:
Fixes which are still left also include-
- [ ] Start button should not work for the joiner
- [ ] Fixing Collaboration traceback
- [ ] Check why clicking on black parts of a card is ignored
- [ ] Fix the issue: after winning a path, pressing the play bar incremented the score repeatedly

quozl commented:
By the way, I've updated the checkboxes in previous comments here.

chimosky commented on PathsActivity.py:
```diff
@@ -84,7 +84,17 @@ def __init__(self, handle):
-        self._setup_presence_service()
+        #self._setup_presence_service()
+        self.connect('shared', self._shared_cb)
+        self.connect('joined', self._joined_cb)
```
You can remove this line as you connected the joined signal in collabwrapper to the same callback.

quozl replied:
Signals mean different things, even though they are both called "joined".  It is best to use the signals and API at the same level of the stack.

chimosky replied:
@quozl could you clarify what you meant by "It's best to use the signals and API at the same level of the stack" as they're both different signals and both get emitted differently.

My statement above was because as `CollabWrapper` is used for collaboration, we should just call the `joined` event of collabwrapper and not the one from the `sugar3.activity.activity.Activity` class.

quozl replied:
> @quozl could you clarify what you meant by "It's best to use the signals and API at the same level of the stack" as they're both different signals and both get emitted differently.

CollabWrapper is a software layer, or stack level, above the Telepathy API exposed by the sugar3 classes.  So whenever someone mixes signals and API between Telepathy API and CollabWrapper, there's an opportunity to bypass a layer, which makes code non-modular and confusing.

> My statement above was because as `CollabWrapper` is used for collaboration, we should just call the `joined` event of collabwrapper and not the one from the `sugar3.activity.activity.Activity` class.

Yes, that's sort of the same thing, except I gave reasons?  :grin:

JuiP committed:
- Remove unnecessary imports and files
- Change unused local variable

JuiP committed:
- revert the change the if condition of _game_over
- Fix method call
- correct the repository in activity/activity.info
